iridauploader/core/upload.py

```diff
@@ -0,0 +1,253 @@
+
+    upload_list = []
+    delayed_list = []
+    for directory_status in directory_status_list:
+        logging.info("Analysing directory: {}".format(directory_status.directory))
+        # ignore invalid directories
+        if directory_status.status_equals(DirectoryStatus.INVALID):
+            continue
+        # Check if run is new, if delay config is > 0, delay the run, else add to upload list
+        elif directory_status.status_equals(DirectoryStatus.NEW):
+            if delay_minutes > 0:
+                upload_helpers.set_run_delayed(directory_status)
+                logging.info("Run has been delayed for {} minutes.".format(delay_minutes))
+                delayed_list.append(directory_status)
+            else:
+                upload_list.append(directory_status)
+        # If run was delayed, check if run can now be uploaded, else it's still delayed
+        elif directory_status.status_equals(DirectoryStatus.DELAYED):
+            if upload_helpers.delayed_time_has_passed(directory_status, delay_minutes):
+                logging.info("Delayed run is ready for upload. Continuing...")
+                upload_list.append(directory_status)
+            else:
+                logging.info("Delayed run is still not ready for upload.")
+                delayed_list.append(directory_status)
+        # other statuses are error, partial, and complete runs, force upload allows theses
+        elif force_upload:
+            upload_list.append(directory_status)
+
+    # Display delayed run count to the user
+    if len(delayed_list) > 0:
+        logging.info("{} run(s) have been delayed.".format(len(delayed_list)))
+
+    # Display runs to upload count to the user
+    if force_upload:
+        logging.info("Starting upload for all non invalid and non delayed runs. {} run(s) found. "
+                     "(Running with --force)".format(len(upload_list)))
+    else:
+        logging.info("Starting upload for all new runs. {} run(s) found.".format(len(upload_list)))
+
+    # get default upload mode if None
+    if upload_mode is None:
+        upload_mode = api_handler.get_default_upload_mode()
+
+    # run upload, keep track of which directories did not upload
+    error_list = []
+    for directory_status in upload_list:
+        logging.info("Starting upload for {}".format(directory_status.directory))
+        result = _validate_and_upload(directory_status, upload_mode)
+        if result.exit_code == exit_return.EXIT_CODE_ERROR:
+            error_list.append(directory_status)
+
+    logging.info("Uploads completed with {} error(s)".format(len(error_list)))
+    for directory_status in error_list:
+        logging.warning("Directory '{}' upload exited with ERROR, check log and status file for details"
+                        "".format(directory_status.directory))
+
+    logging.info("Batch upload complete, Exiting!")
+    return exit_success()
+
+
+def _validate_and_upload(directory_status, upload_mode):
```

Comment:
Big fan of how much more concise this method is.  Thanks for the refactor here.

Reply:
No problem, I wish I had taken a picture of my whiteboard while I was figuring out how to redo the code flow and exception handling for it.